### PR TITLE
Extend C Api to fetch the layout and ndim from External Source

### PR DESCRIPTION
**Pull request by banasraf**
Signed-off-by: Rafal <Banas.Rafal97@gmail.com>

This PR add C Api calls (and corresponding Pipeline api calls) to list the external inputs in the pipeline and fetch their ndim and layout.

## Category:

New feature

## Description:

Fetching the ndim and layout is needed to implement auto-configuration in DALI Triton backend. 

### Affected modules and functionalities:

C Api, Pipeline

### Key points relevant for the review:

Pipeline API implementations

### Tests
- [ ] Existing tests apply
- [x] New tests added
  - [ ] Python tests
  - [x] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [x] Documentation updated
  - [x] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-2748

### CODE REVIEW

**Comment by banasraf** [comment]
!build

**Review by awolant — APPROVED** [review]

File: include/dali/c_api.h [hunk 1]
```diff
@@ -276,6 +276,48 @@ daliSetExternalInputTensors(daliPipelineHandle *pipe_handle, const char *name,
...
+ * @param pipe_handle Pointer to pipeline handle.
+ * @param name Name of the external input.
+ * @return Layout string.
+ */
+DLL_PUBLIC const char *daliGetExternalInputLayout(daliPipelineHandle *pipe_handle,
+                                                  const char *name);
+
+/**
+ * @brief Get the data dimensionality required by the external input with a given name.
+ * If the dimensionality is not determined, -1 is returned.
```
[the remark this answers is not in the document]
Comment (banasraf) [reply include/dali/c_api.h]: done

File: dali/pipeline/pipeline.h [hunk 2]
```diff
@@ -112,12 +113,17 @@ class DLL_PUBLIC Pipeline {
    * Equivalent to inserting ExternalSource with output of given name and specified
    * device placemnt.
    */
-  DLL_PUBLIC inline int AddExternalInput(const string &name, const string &device = "cpu") {
-    return AddOperator(OpSpec("ExternalSource")
-                           .AddArg("name", name)
-                           .AddArg("device", device)
-                           .AddOutput(name, device),
-                       name);
+  DLL_PUBLIC inline int AddExternalInput(const string &name,
```
[the remark this answers is not in the document]
Comment (banasraf) [reply dali/pipeline/pipeline.h]: done

File: dali/pipeline/pipeline.cc [hunk 3]
```diff
@@ -754,6 +757,50 @@ ReaderMeta Pipeline::GetReaderMeta(std::string name) {
...
+const std::string &Pipeline::input_name(int n) const {
+  DALI_ENFORCE(n >= 0, "Id of an external source must be a non-negative integer.");
+  for (const auto &name : input_names_) {
+    if (n == 0) {
+      return name;
+    }
+    --n;
+  }
+  DALI_FAIL(make_string("Trying to fetch the name of the ", n, ". external input while "
+                        "the pipeline has only ", num_inputs(), " inputs."));
```
[the remark this answers is not in the document]
Comment (banasraf) [reply dali/pipeline/pipeline.cc]: Good catch, fixed

**Review by szalpal — APPROVED** [review]

**Comment by banasraf** [comment]
!build

**Pull request merged**